docs/understanding-airbyte/airbyte-protocol.md

```diff
@@ -0,0 +1,735 @@
+
+### Source Sync Modes
+* `incremental` - send all the data for the Stream since the last sync (e.g. the state message passed to the Source). This is the most common sync mode. It only sends new data.
+* `full_refresh` - resend all data for the Stream on every sync. Ignores State. Should only be used in cases where data is very small, there is no way to keep a cursor into the data, or it is necessary to capture a snapshot in time of the whole dataset. Be careful using this, because misuse can lead to sending much more data than expected.
+
+### Destination Sync Modes
+* `append` - add new data from the sync to the end of whatever already data already exists.
+* `append_dedup` - add new data from the sync to the end of whatever already data already exists and deduplicate it on primary key. This is the most **common** sync mode. It does require that a primary exists in the data. This is also known as SCD Type 1 & 2.
+* `overwrite` - replace whatever data exists in the destination data store with the data that arrives in this sync.
+
+## Logic for resolving the Cursor Field
+
+This section lays out how a cursor field is determined in the case of a Stream that is doing an `incremental` sync.
+
+* If `source_defined_cursor` in `AirbyteStream` is true, then the source determines the cursor field internally. It cannot be overridden. If it is false, continue...
+* If `cursor_field` in `ConfiguredAirbyteStream` is set, then the source uses that field as the cursor. If it is not set, continue...
+* If `default_cursor_field` in `AirbyteStream` is set, then the sources use that field as the cursor. If it is not set, continue...
+* Illegal - If `source_defined_cursor`, `cursor_field`, and `default_cursor_field` are all false-y, this is an invalid configuration.
+
+## Schema Mismatches
+
+Over time, it is possible for the catalog to become out of sync with the underlying data store it represents. The Protocol is design to be resilient to this. In should never fail due to a mismatch.
+
+|Scenario|Outcome|
+|---|---|
+|Stream exists in catalog but not in data store|When the source runs, it will not find the data for that stream. All other streams sync as usual.|
+|Stream exists in data store but not in catalog|When the source runs, it never looks for the data in the store related to that stream and thus does not emit it.
+|Field exists in catalog but not in data store|If the column for a table is remove in the underlying data store the Source will not find it and will not replicate it. It should not cause a failure. The data simply will not be there.|
+|Field exists in data store but not in catalog|When the source runs, it never looks for the field in the store. It should not emit that field. If it does, it should be ignored downstream. The existence of an unknown field should not cause a failure.|
+
+In short, if the catalog is ever out of sync with the schema of the underlying data store, it should never block replication for data that is present.
```

Comment:
Above we say the following, which I don't think matches the intent of this section.

> The destination should fail if any of the messages it receives do not match the structure described in the `catalog`.

Should the destination `exit 1` or ignore messages that are out-of-sync with the spec?

Reply:
good call out. ignore.

Reply:
updated above:
> * The destination should ignore fields or streams that are out of sync with the `catalog`. The destination should always make its best effort to load what data is there that does match that catalog. e.g. if the User Stream has the fields first_name and last_name in the catalog, but the record has first_name and eye_color, the destination should persist first_name, even though last_name is missing. It should ignore eye_color as extraneous.

Reply:
is this section specifically referring to sources?